Effect aliasing

Orbion-J opened the pull request:
I added the possibility to write effect aliases, similarly as what already exists for types.
- there is a new keyword `effectname`
- to write an alias for an effect row : `effectname MyEffectRow(a, ... ,e::Eff, ...) = { Op1 : type, ... | e }` the arguments being type variable of any kind (kinds other than type must be explicit)
- we can have, as above, an open row (the row variable is a parameter of the effect alias) or a closed one (just do not write the `| e`
- To use it in a signature or in another type or effect alias just apply it with the right arguments as for `typename` things.
- In arrows, we can use them as row variables: `() -MyEffectRow(args)-> ()` (idem with `~>`)
- However, due to lack of kind inference, row variables and aliases have to be used carefully so that links does not think they are of kind type. We need to write them most of the time between braces ` { | ... }`. For instance, if you have `effectname E(a::Eff) = {X : ... | a }` and a row variable `e::Eff`, you will have to write `E({ |e})`. (Idem for another effect alias instead of the variable). This makes the usage of several nested aliases a bit messy, it would be nice if we could avoid it.
- We cannot write recursive effect aliases for now. In the branch `visitor`, I added another transformer that makes possible simple recursion by inlining a mu type in one pass.
- For now the aliases are replaced by the row they correspond to : we do not keep aliases. 
- In the repl, effect alias definitions are printed but without the braces ! Rows are in general printed without braces and the alias body is a row. => this might need to be enhanced

About implementation, I copied and then merged most of the time what existed for `typename`.

jamescheney commented:
I don't know the backstory of this, and haven't had a chance to look at this carefully, but instead of adding a keyword and duplicating a lot of code for `typename`, would it be possible (and wouldn't it be nicer) to augment `typename` to allow an optional kind, and generalize it to allow any kind of type alias (with Type being the default used if no explicit kind is given)?

e.g. instead of 
```
effectname MyEffectRow(a, ... ,e::Eff, ...) = { Op1 : type, ... | e }
```
could we write
```
typename MyAlias(...) :: K = ty
```
for any kind K, with Eff being a special case, and just generalize the existing code for aliases to allow this?

The description and comments suggest that this may not be straightforward, because of existing idiosyncrasies, but the above design seems cleaner and something we should work toward (see also #44 and comments about row aliases).  Perhaps a constructive compromise could be to adopt this PR with a small change: allowing (in addition or only)
```
typename  MyEffectRow(a, ... ,e::Eff, ...):: Eff =  { Op1 : type, ... | e }
```
as a special case without (yet) trying to generalize to allow aliases of any kind.

Orbion-J commented:
@jamescheney I can surely do something that looks like that. At least syntacticly, I can check during parsing which kind do we have and use after the same things I ve already done.
I will still try see how complicated it would be to extend to presence, in order to be able to at least use it for every primary kind.

dhil commented on bin/repl.ml:
```diff
@@ -155,14 +155,14 @@ let rec directives : (string * ((Context.t -> string list -> Context.t) * string
                 (Module_hacks.Name.prettify k)
                 (Types.string_of_tycon_spec s))
-            (StringSet.diff (Env.String.domain tycon_env) (Env.String.domain Lib.typing_env.Types.tycon_env));
+            (StringSet.diff (Env.String.domain tycon_env) (Env.String.domain Lib.typing_env.Types.alias_env));
```
```suggestion
            (StringSet.diff (Env.String.domain tycon_env) (Env.String.domain Lib.typing_env.Types.tycon_env));
```

dhil commented on bin/repl.ml:
```diff
@@ -155,14 +155,14 @@ let rec directives : (string * ((Context.t -> string list -> Context.t) * string
           let tycon_env =
             let tenv = Context.typing_environment context in
-            tenv.Types.tycon_env
+            tenv.Types.alias_env
```
```suggestion
            tenv.Types.tycon_env
```

dhil commented on bin/repl.ml:
```diff
@@ -221,7 +221,7 @@ let rec directives : (string * ((Context.t -> string list -> Context.t) * string
           let tenv, aliases =
             let tyenv = Context.typing_environment context in
-            tyenv.Types.var_env, tyenv.Types.tycon_env
+            tyenv.Types.var_env, tyenv.Types.alias_env
```
```suggestion
            tyenv.Types.var_env, tyenv.Types.tycon_env
```

dhil commented on bin/repl.ml:
```diff
@@ -268,8 +268,8 @@ let handle previous_context current_context = function
        let tenv' = Context.typing_environment current_context in
-       let tycon_env, tycon_env' =
-         Types.(tenv.tycon_env, tenv'.tycon_env)
+       let tycon_env, tycon_env'=
```
```suggestion
       let tycon_env, tycon_env' =
```

dhil commented on bin/repl.ml:
```diff
@@ -268,8 +268,8 @@ let handle previous_context current_context = function
-       let tycon_env, tycon_env' =
-         Types.(tenv.tycon_env, tenv'.tycon_env)
+       let tycon_env, tycon_env'=
+         Types.(tenv.alias_env, tenv'.alias_env)
```
```suggestion
         Types.(tenv.tycon_env, tenv'.tycon_env)
```

dhil commented on core/defaultAliases.ml:
```diff
@@ -3,7 +3,7 @@ open CommonTypes
 module AliasEnv = Env.String
 
-let alias_env : Types.tycon_environment =
+let alias_env : Types.alias_environment =
```
```suggestion
let alias_env : Types.tycon_environment =
```

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -68,7 +72,7 @@ module Desugar = struct
     List.map SugarQuantifier.get_resolved_exn sqs
 
-  let rec datatype (alias_env : Types.tycon_environment) t' =
+  let rec datatype (alias_env : Types.alias_environment) t' =
```
```suggestion
  let rec datatype (alias_env : Types.tycon_environment) t' =
```

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -209,6 +213,72 @@ module Desugar = struct
+                       let  t = match_kinds i (q, t) in
+                       type_arg alias_env t node)
+              in
+
```
```suggestion
```

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -209,6 +213,72 @@ module Desugar = struct
+                let t_kind = primary_kind_of_type_arg t in
+                if q_kind <> t_kind then
+                  raise
+                    (TypeApplicationKindMismatch
```
`TypeApplicationKindMismatch` should probably be defined in `errors.ml`.

Orbion-J replied:
yes it is, what should I do there ?

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -209,6 +213,72 @@ module Desugar = struct
+                raise (TypeApplicationArityMismatch { pos = node.pos; name = name; expected = qn; provided = tn })
+            in
+            begin match SEnv.find_opt name alias_env with
+              | None -> raise (UnboundTyCon (node.pos, name))
```
`UnboundTyCon` should probably also be defined in `errors.ml`

Orbion-J replied:
idem

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -236,11 +306,16 @@ module Desugar = struct
     (dt, Some (datatype alias_env dt))
 
+  let row' alias_env ((r, _) :row') =
+    (r, Some (row alias_env r (WithPos.make (Datatype.Effect r)))) (* TODO(rj) should keep the pos *)
```
I think this is fine for now. Ideally, `row` would be part of the datatype definition -- we should work towards that at a later point.

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -340,36 +417,43 @@ object (self)
          * environment, as mutuals. Quantifiers need to be desugared. *)
-        let ((mutual_env : tycon_spec SEnv.t), ts) =
-          List.fold_left (fun (alias_env, ts) {node=(t, args, (d, _)); pos} ->
+        let ((mutual_env : Types.alias_environment), ts) =
```
```suggestion
        let ((mutual_env : Types.tycon_environment), ts) =
```

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -340,36 +417,43 @@ object (self)
+            let b' = match Desugar.aliasbody mutual_env b with
+                | Typename   (_, Some _) as b' -> b'
+                | Effectname (_, Some _) as b' -> b'
+                | _ -> assert false
```
You should use `internal_error` here.

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -468,17 +557,17 @@ let toplevel_bindings alias_env bs =
 let sentence typing_env = function
   | Definitions bs ->
-      let _alias_env, bs' = toplevel_bindings typing_env.tycon_env bs in
+      let _alias_env, bs' = toplevel_bindings typing_env.alias_env bs in
```
`tycon_env`

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -468,17 +557,17 @@ let toplevel_bindings alias_env bs =
+      let _alias_env, bs' = toplevel_bindings typing_env.alias_env bs in
         Definitions bs'
-  | Expression  p  -> let _o, p = phrase typing_env.tycon_env p in
+  | Expression  p  -> let _o, p = phrase typing_env.alias_env p in
```
`tycon_env`

dhil commented on core/desugarDatatypes.ml:
```diff
@@ -468,17 +557,17 @@ let toplevel_bindings alias_env bs =
 let program typing_env (bindings, p : Sugartypes.program) :
     Sugartypes.program =
-  let alias_env = typing_env.tycon_env in
+  let alias_env = typing_env.alias_env in
```
`tycon_env`

dhil commented on core/desugarDatatypes.mli:
```diff
@@ -1,4 +1,4 @@
-val read : aliases:Types.tycon_environment -> string -> Types.datatype
+val read : aliases:Types.alias_environment -> string -> Types.datatype
```
```suggestion
val read : aliases:Types.tycon_environment -> string -> Types.datatype
```

dhil commented on core/desugarEffects.ml:
```diff
@@ -1089,16 +1091,28 @@ class main_traversal simple_tycon_env =
+                        StringMap.add t (StringSet.elements used_mutuals) dep_graph
+                    in
+                    (implicits, dep_graph)
+                  | Effectname (r,_) -> (* is this the right thing to do ? *)
```
We have to be careful here. I am not sure this is the right thing to do. We need to check what the effect desugaring procedure is doing under the hood. The arrows on labels in effect rows means something else than function arrows.

Orbion-J replied:
yes, I am not really sure of what this is really doing. I tried a bit to understand what happens but I did not really manage to do it...
There is something else we may want to do there. For now, effectname does not support a useful syntactic sugar. We have this :
```ocaml
links> effectname E = { Get:Int } ;
E = Get:Int
```
But we might want this ;
```ocaml
links> effectname E = { Get:Int } ;
E = Get:() {}-> Int
```
It should happens in this file but I did not figured out what to do to make this work yet.

dhil replied:
The latter is the correct behaviour, I believe.

Orbion-J replied:
it works now, with a bit of a hack
I added something also we might want (or not) :
```ocaml
typename F = () {}-> Int ;
effectname E = {Get:F} ;
effectname E' = {Get':Int} ;
```
now, we get `E' = Get': () {}-> Int` and also `E = Get:() {}-> Int`. (before we would have had `E = Get:() {}-> () {}-> Int`) 
If we detect a function type application for an operation, it is not desugared into a new function type with unit input. I think this is what we might want

dhil commented on core/parser.mly:
```diff
@@ -1130,7 +1136,7 @@ type_arg_list:
 | datatype                                                     { Datatype.Type $1     }
 | braced_fieldspec                                             { Datatype.Presence $1 }
-| LBRACE row RBRACE                                            { Datatype.Row $2      }
+| LBRACE trow RBRACE                                           { Datatype.Row $2      }
```
I think this should just be `row`.

Orbion-J replied:
I just merged `trow` and `erow`.
Initially, I did this because `efields` are used in arrows so it seemed to me the best thing to use. Do you really think we should add the effect application to the `row` ? And what about the dot thing ? I think that, when it is activated, we should be able to use the dot in effect aliases, isn't it ?

dhil commented on core/parser.mly:
```diff
@@ -1185,21 +1199,32 @@ vfield:
 | CONSTRUCTOR                                                  { ($1, present) }
 | CONSTRUCTOR fieldspec                                        { ($1, $2)      }
 
+tfields:
```
The `fields_def` macro is supposed to eliminate some of the redundancy w.r.t. rows. We should use that rather than duplicating the parsing rules for rows.

Orbion-J replied:
cf supra

dhil commented:
> I don't know the backstory of this, and haven't had a chance to look at this carefully, but instead of adding a keyword and duplicating a lot of code for `typename`, would it be possible (and wouldn't it be nicer) to augment `typename` to allow an optional kind, and generalize it to allow any kind of type alias (with Type being the default used if no explicit kind is given)?

Ultimately, we want to do something like that. But we before we do that we should have a design in mind. The point of this change is to introduce the minimal thing that let us tweak the programming experience / presentation with / of effects and handlers.

In its current state this patch is more invasive than I had originally intended. However, it is probably fine, but we need to be a bit more careful about mixing kind type and row in one alias environment -- I am sure this approach opens up an avenue for a lot of new bugs.

Orbion-J commented:
I fixed the simple things.
I tried to add what james was suggesting, I had to do a a bit of a hack in the parser to make it work but it seems to be ok. I added also a few things for presence but this is neither complete nor working.
Anyway, you are right mixing everything could cause some unexpected behavior as :
```ocaml
links> typename A = Int ;
A = Int
links> effectname B = { Op:() -A-> () } ;
***: Error: File "core/types.ml", line 1435, characters 11-17: Assertion failed 
```
this should for instance cause a kind mismatch. I think checking in `instantiate.ml` would help but there could be other issues.

Orbion-J commented on core/parser.mly:
```diff
@@ -413,6 +430,7 @@ arg:
 | UFLOAT                                                       { string_of_float' $1 }
 | TRUE                                                         { "true"  }
 | FALSE                                                        { "false" }
+| DEFAULT                                                      { "default" }
```
this was missing, making impossible to reset settings to default since
```ocaml
links> @set something default ;
```
caused a parsing error

dhil replied:
You should submit this in a separate PR, do not include it in this one.

Orbion-J replied:
ok, i ll do that

dhil commented:
> I fixed the simple things. I tried to add what james was suggesting, I had to do a a bit of a hack in the parser to make it work but it seems to be ok. I added also a few things for presence but this is neither complete nor working. Anyway, you are right mixing everything could cause some unexpected behavior as :
> 
> ```ocaml
> links> typename A = Int ;
> A = Int
> links> effectname B = { Op:() -A-> () } ;
> ***: Error: File "core/types.ml", line 1435, characters 11-17: Assertion failed 
> ```
> 
> this should for instance cause a kind mismatch. I think checking in `instantiate.ml` would help but there could be other issues.

Yes. Do not implement James' suggestion. That's for later when we have worked out a proper design. It is too much to bundle it all up.

jamescheney commented:
> Do not implement James' suggestion. That's for later when we have worked out a proper design. It is too much to bundle it all up.

Of course, I was just asking if this was a "good enough for now" change or intended as a long-term proper design.

Orbion-J commented on core/instantiate.ml:
```diff
@@ -169,6 +169,7 @@ let instantiates : instantiation_maps -> (datatype -> datatype) * (row -> row) *
       match t with
       | Row _ -> t
       | Meta row_var -> Row (StringMap.empty, row_var, false)
+      | Alias (_,row) -> row
```
cf https://github.com/links-lang/links/pull/1141#issuecomment-1141321170

dhil commented on core/desugarEffects.ml:
```diff
@@ -978,6 +982,7 @@ class main_traversal simple_tycon_env =
       let module D = Datatype in
       let o, rv =
         match rv with
+        | D.EffectApplication _ -> (o, rv) (* TODO(rj) do i need to do something there ? *)
```
You need to make sure the alias constructor is desugared and then make sure the arguments are desugared.

dhil commented on core/types.ml:
```diff
@@ -1429,7 +1428,12 @@ and flatten_row : row -> row = fun row ->
     (* HACK: this probably shouldn't happen! *)
     | Meta row_var -> Row (StringMap.empty, row_var, false)
-    | _ -> assert false in
+    (* | Alias (PrimaryKind.Row, _, row) -> row *)
```
Maybe remove this?

dhil commented on core/types.ml:
Maybe we should raise `internal_error` here?

dhil reviewed (changes requested):
LGTM modulo the requested changes.

dhil reviewed (approved):
LGTM. I will merge this tomorrow.
